### DISCUSSION TITLE
ZHA Device Handlers version 2 (ZHA-Quirks V2) Zigbee device entities architectural design developer discussion

### DISCUSSION THREAD

**Discussion by Hedda** [home-assistant/architecture#1032] [General]
FYI, just want to raise awareness for advanced users and potential developers of the built-in [ZHA (Zigbee Home Automation) integration](https://www.home-assistant.io/integrations/zha) following Home Assistant architecture discussion that you/they might be interested in following this related upstream architectural proposal and development discussion about changing how the [zigpy library](https://github.com/zigpy/zigpy) that the ZHA integration depend use [ZHA Device Handlers (a.k.a. quirks)](https://github.com/zigpy/zha-device-handlers) to decouple entities and expose them in the ZHA integration from underlying Zigbee devices, clusters, and endpoints to potentially make it easier to develop quirks as ZHA Device Handlers without know everything about the official ZCL (Zigbee Cluster Library) specifications:

https://github.com/zigpy/zigpy/discussions/1312

So if interested please check out this RFC and consider contributing your constructive feedback/input there.

Summary; the developer discussion there includes the idea of introducing a new API concept that will allow ZHA-quirks developers to add additional custom entities support for a new device that uses non-standard Zigbee clusters and attributes without having to both create a [Zigbee device handler "quirk" in the ZHA Device Handlers (zha-quirks) repository](https://github.com/zigpy/zha-device-handlers/) and modify both the [codebase of the ZHA component in the Home Assistant code](https://github.com/home-assistant/core/tree/dev/homeassistant/components/zha). 

The new proposed concept was raised now by [puddly](https://github.com/puddly) with a suggestion on how to make quirks easier and less complicated to create and edit by new ZHA-quirks developers, but does add further Home Assisstant integrations in the upstream library that is also used by other third-party projects that are not using the ZHA component.

For high-level end-user explanation of why ZHA Device Handlers (a.k.a. quirks) are needed suggest read these two sections in ZHA docs:

https://www.home-assistant.io/integrations/zha#knowing-which-devices-are-supported

https://www.home-assistant.io/integrations/zha#how-to-add-support-for-new-and-unsupported-devices

PS: Also reference:

https://github.com/zigpy/zha-device-handlers/blob/dev/README.md

**_ZHA Device Handlers are custom quirks implementations for [Zigpy](https://github.com/zigpy/zigpy), the library that provides the [Zigbee](http://www.zigbee.org/) support for the [ZHA](https://www.home-assistant.io/components/zha/) component in [Home Assistant](https://www.home-assistant.io/)._**

**_ZHA device handlers bridge the functionality gap created when manufacturers deviate from the ZCL specification, handling deviations and exceptions by parsing custom messages to and from Zigbee devices. Zigbee devices that deviate from or do not fully conform to the standard specifications set by the Zigbee Alliance may require the development of custom ZHA Device Handlers (ZHA custom quirks handler implementation) to for all their functions to work properly with the ZHA component in Home Assistant._**

**_Custom quirks implementations for zigpy implemented as ZHA Device Handlers are a similar concept to that of [Hub-connected Device Handlers for the SmartThings Classics platform](https://stdavedemo.readthedocs.io/en/latest/device-type-developers-guide/) as well that of [Zigbee-Herdsman Converters (formerly Zigbee-Shepherd Converters) as used by Zigbee2mqtt](https://www.zigbee2mqtt.io/advanced/support-new-devices/01_support_new_devices.html), meaning they are virtual representation of a physical device that expose additional functionality that is not provided out-of-the-box by the existing integration between these platforms. See [Device Specifics](https://github.com/zigpy/zha-device-handlers/blob/dev/README.md#Device-Specifics) for details._**

**Comment by frenck** [comment]
> just want to raise awareness

Closing this, as this is not an implementation proposal you are implementing.

This place is not meant as a forum.

../Frenck
